edgeless_api/proto/messages.proto

```diff
@@ -63,6 +63,54 @@ message SpawnFunctionResponse {
     optional InstanceId instance_id = 2;
 }
 
+// Possible message types of FunctionInstance::UpdateNode().
+enum UpdateNodeRequestType {
+    REGISTER   = 0;
+    DEREGISTER = 1;
+}
+
+// Request message of FunctionInstance::UpdateNode().
+message UpdateNodeRequest {
+    // Request type: registration or deregistration. Always present.
+    UpdateNodeRequestType request_type = 1;
+    // Identifier of the node that is entering/leaving the orchestration domain.
+    // This field must be always present.
+    string node_id                     = 2;
+    // URL of the node's agent server. Always present if REGISTER type.
+    optional string agent_url          = 3;
+    // URL of the node's invocation server. Always present if REGISTER type.
+    optional string invocation_url     = 4;
+}
+
+// Response message of FunctionInstance::UpdateNode().
+message UpdateNodeResponse {
+    // If present it means that the request has been rejected.
+    // In this case, other fields may not be present or contain meaningful data.
+    optional ResponseError response_error = 1;
+}
+
+// Possible message type of FunctionInstance::UpdatePeersRequest().
+enum UpdatePeersRequestType {
+    ADD   = 0; // add one peer
+    DEL   = 1; // delete one peer
+    CLEAR = 2; // delete all peers
+}
+
+// Request message of FunctionInstance::UpdatePeersRequest().
+message UpdatePeersRequest {
+    // Request type. Always present.
+    UpdatePeersRequestType request_type = 1;
+    // Identifier of the peer node. Always present, unless type is CLEAR.
+    optional string node_id             = 2;
+    // URL of the peer node's invocation server. Always present if ADD type.
```

Comment:
This is very personal. The alternative would be to have dedicated Register vs. Deregister messages, where all fields are mandatory.

The latter might have a little higher maintenance costs:

1. with the single-message approach you do one test and it covers everything
2. if you ever want to add new fields, then it's a matter of extending just one test

Since I have a telco background, the presence of optional fields does not struck me as odd (just take a look to a random 3GPP specs doc...) but I see your point.